bug in `list_tfrecord_files`

When passing a single file to `build_dataset`, the following error occurs:

```
def list_tfrecord_files(
        files: Union[List[os.PathLike], os.PathLike],
    ) -> List[os.PathLike]:
        """Parse the input into the list of files ending with '.tfrecord'."""
        if not isinstance(files, list):
            fn, ext = os.path.splitext(files)
            if ext != "tfrecord":
                pth = Path(files)
>               files = [pth / f for f in os.listdir(files) if f.endswith(".tfrecord")]
E               NotADirectoryError: [Errno 20] Not a directory: '/tmp/pytest-of-quantumjot/pytest-15/test_read_write_dtype_uint16_0/test.tfrecord'
```

@KristinaUlicna @laureho

--- a/_tests/test_dataset.py
+++ b/_tests/test_dataset.py
@@ -4,7 +4,12 @@ from pathlib import Path
 import numpy as np
 import pytest
 
-from cellx.tools.dataset import build_dataset, write_dataset
+from cellx.tools.dataset import (
+    build_dataset,
+    count_images_in_dataset,
+    list_tfrecord_files,
+    write_dataset,
+)
 
 
 def _make_images(n_images: int = 10, shape: tuple = (64, 64, 1)):
@@ -13,26 +18,28 @@ def _make_images(n_images: int = 10, shape: tuple = (64, 64, 1)):
     return rng.integers(0, 255, size=(n_images,) + shape).astype(np.uint8)
 
 
+def _test_write_images(filename, dtype="uint8"):
+    """Make some images and write them out."""
+    images = _make_images().astype(dtype)
+    write_dataset(filename, images)
+    return images
+
+
 def test_write_dataset(tmp_path):
     """Test writing out a dataset."""
-    images = _make_images()
-
-    # note(arl) - this function doesn't work with a path as input!!
     filename = Path(tmp_path) / "test.tfrecord"
-    write_dataset(str(filename), images)
+    _ = _test_write_images(filename)
 
     # make sure the exported filename exists
     assert os.path.exists(filename)
 
 
 @pytest.mark.parametrize("dtype", ["uint8", "uint16"])
-def test_read_write_dtype(tmp_path, dtype):
+def test_read_write_dataset_dtype(tmp_path, dtype):
     """Test to make sure serialized data matches raw data."""
-    images = _make_images().astype(dtype)
-
     # note(arl) - this function doesn't work with a path as input!!
     filename = Path(tmp_path) / "test.tfrecord"
-    write_dataset(str(filename), images)
+    images = _test_write_images(filename, dtype=dtype)
 
     # build the dataset and return the first record
     ds = build_dataset(tmp_path, output_dtype=dtype)
@@ -43,3 +50,39 @@ def test_read_write_dtype(tmp_path, dtype):
 
     # make sure the data matches
     np.testing.assert_equal(x, images[0, ...])
+
+
+def test_list_tfrecords(tmp_path):
+    """Check that list_tfrecords returns a list of tfrecord files when using
+    different inputs, e.g. single file, folder or list."""
+    filename = Path(tmp_path) / "test.tfrecord"
+    _ = _test_write_images(filename)
+
+    # single path to file
+    files = list_tfrecord_files(filename)
+    assert files == [filename]
+
+    # list of files
+    files = list_tfrecord_files([filename])
+    assert files == [filename]
+
+    # path as string
+    files = list_tfrecord_files(tmp_path)
+    assert files == [filename]
+
+    # path as `Path`
+    files = list_tfrecord_files(Path(tmp_path))
+    assert files == [filename]
+
+
+@pytest.mark.parametrize("n_images", [1, 10, 100])
+def test_count_images(tmp_path, n_images):
+    """Make datasets with varying numbers of images. Test counting the number
+    of images in the dataset."""
+    filename = Path(tmp_path) / "test.tfrecord"
+    images = _make_images(n_images=n_images)
+    write_dataset(filename, images)
+
+    # count the number of images in the dataset
+    n = count_images_in_dataset(filename)
+    assert n == n_images

--- a/cellx/tools/dataset.py
+++ b/cellx/tools/dataset.py
@@ -22,7 +22,9 @@ def _float_feature(value):
 
 
 def write_dataset(
-    filename: str, images: np.ndarray, labels: Optional[np.ndarray] = None
+    filename: Union[str, os.PathLike],
+    images: np.ndarray,
+    labels: Optional[np.ndarray] = None,
 ):
     """Write out a TF record of image data for training models.
 
@@ -36,8 +38,11 @@ def write_dataset(
         An array of labels.
     """
 
-    if not filename.endswith(".tfrecord"):
-        filename = f"{filename}.tfrecord"
+    # convert to a Path, although TF likes a string as input
+    filename = Path(filename)
+    if filename.suffix != ".tfrecord":
+        filename = filename.with_suffix(".tfrecord")
+    filename = str(filename)
 
     assert images.dtype in (
         np.uint8,
@@ -143,12 +148,25 @@ def per_channel_normalize(x: tf.Tensor) -> tf.Tensor:
 def list_tfrecord_files(
     files: Union[List[os.PathLike], os.PathLike],
 ) -> List[os.PathLike]:
-    """Parse the input into the list of files ending with '.tfrecord'."""
+    """Parse the input into the list of files ending with '.tfrecord'.
+
+    Parameters
+    ----------
+    files : List[os.PathLike] or os.PathLike
+        Parse a list or single pathlike objects
+
+    Returns
+    -------
+    files : List[os.PathLike]
+        A TF compatible list of paths to `.tfrecord` files.
+
+    """
     if not isinstance(files, list):
-        fn, ext = os.path.splitext(files)
-        if ext != "tfrecord":
-            pth = Path(files)
-            files = [pth / f for f in os.listdir(files) if f.endswith(".tfrecord")]
+        files = Path(files)
+        if files.is_dir():
+            files = [f for f in files.iterdir() if f.suffix == ".tfrecord"]
+        else:
+            files = [files]
     return files
 
 

